### ISSUE TITLE
[Tosa backend]Generated tosa ir has a semantic error

### ISSUE THREAD

**Issue by JiaweiRen2022** [llvm/torch-mlir#2581]
I wrote a matmul operation using PyTorch and lowering it to the Tosa dialect using torch-mlir. However, the generated Tosa dialect contains semantic errors and cannot be further lowering .

## test case
```
class Matmul(nn.Module):
    def __init__(self):
        super(Matmul, self).__init__()
 
    def forward(self, x, weight):
        r = torch.matmul(x, weight)
        return r

def gen_mt_mlir():
    model = Matmul()
    weight = torch.randn(80,300,250,150,dtype=torch.float32)
    input = torch.ones(300,100,250)
    res = model(input,weight)
    print(res.shape)
    module = torch_mlir.compile(model, [input,weight], output_type="TOSA")

if __name__=='__main__':
    gen_mt_mlir()
```

## result analysis
generated tosa dialect ir is as follows:
```
module attributes {torch.debug_module_name = "Matmul"} {
  func.func @forward(%arg0: tensor<300x100x250xf32>, %arg1: tensor<80x300x250x150xf32>) -> tensor<80x300x100x150xf32> {
    %0 = "tosa.const"() {value = dense<[1, 2, 0, 3]> : tensor<4xi32>} : () -> tensor<4xi32>
    %1 = "tosa.transpose"(%arg1, %0) : (tensor<80x300x250x150xf32>, tensor<4xi32>) -> tensor<300x250x80x150xf32>
    %2 = "tosa.reshape"(%1) {new_shape = array<i64: 300, 250, 12000>} : (tensor<300x250x80x150xf32>) -> tensor<300x250x12000xf32>
    %3 = "tosa.matmul"(%arg0, %2) : (tensor<300x100x250xf32>, tensor<300x250x12000xf32>) -> tensor<300x100x12000xf32>
    %4 = "tosa.reshape"(%3) {new_shape = array<i64: 300, 100, 80, 150>} : (tensor<300x100x12000xf32>) -> tensor<300x100x80x150xf32>
    %5 = "tosa.transpose"(%4, %0) : (tensor<300x100x80x150xf32>, tensor<4xi32>) -> tensor<80x300x100x150xf32>
    return %5 : tensor<80x300x100x150xf32>
  }
}
```
the result tensor shape of `%5 = "tosa.transpose"(%4, %0) : (tensor<300x100x80x150xf32>, tensor<4xi32>) -> tensor<80x300x100x150xf32>` is error . Here, it is necessary to create a new constant for dimension transformation instead of reusing 0%.

**Comment by hunterzju** [comment]
It seems that the matmul operand dims does not match, what do you want to get by matmul a 300x100x250 data with a 80x300x250x150? what is the dim of the result?

**Comment by bilibiliGO283** [comment]
# torch output shape
```
